core/domain/html_validation_service.py

```diff
@@ -0,0 +1,890 @@
+
+                            if err_msg:
+                                if err_msg not in err_dict:
+                                    err_dict[err_msg] = [html_string]
+                                elif html_string not in err_dict[err_msg]:
+                                    err_dict[err_msg].append(html_string)
+                    except Exception:
+                        continue
+
+                for tabs in soup.findAll(name='oppia-noninteractive-tabs'):
+                    # This is wrapped in try catch block since any invalid
+                    # content such as int will not be convertible in Beautiful
+                    # Soup format. Also missing attributes will result in error
+                    # when we try to extract them here.
+                    try:
+                        tab_content_json = unescape_html(
+                            tabs['tab_contents-with-value'])
+                        tab_content_list = json.loads(tab_content_json)
+                        for tab_content in tab_content_list:
+                            content_html = tab_content['content']
+                            soup_for_tabs = bs4.BeautifulSoup(
+                                content_html, 'html.parser')
+                            for tag in soup_for_tabs.findAll(name=tag_name):
+                                err_msg = _validate_customization_args_in_tag(
+                                    tag)
+
+                                if err_msg:
+                                    if err_msg not in err_dict:
+                                        err_dict[err_msg] = [html_string]
+                                    elif html_string not in err_dict[err_msg]:
+                                        err_dict[err_msg].append(html_string)
+                    except Exception:
+                        continue
+
+    return err_dict
+
+
+def _validate_customization_args_in_tag(tag):
+    """Validates customization arguments of Rich Text Components in a soup.
+
+    Args:
+        tag: bs4.element.Tag. The html tag to be validated.
+
+    Returns:
+        str. An empty string if the tag has valid attributes otherwise the
+            error message.
+    """
+
+    with open(feconf.RTE_EXTENSIONS_DEFINITIONS_PATH, 'r') as f:
+        rich_text_components = constants.parse_json_from_js(f)
+
+    # Mapping of type to classes defined in objects.py.
+    obj_types_to_obj_classes = {
+        'SanitizedUrl': objects.SanitizedUrl,
+        'unicode': objects.UnicodeString,
+        'MathLatexString': objects.MathLatexString,
+        'Filepath': objects.Filepath,
+        'int': objects.Int,
+        'bool': objects.Boolean,
+        'html': objects.Html
+    }
+    filename_re = r'^[A-Za-z0-9+/]*\.((png)|(jpeg)|(gif)|(jpg))$'
+
+    tag_name = tag.name
+    tag_key = tag_name.replace('oppia-noninteractive-', '').capitalize()
+    component = rich_text_components[tag_key]
```

Comment:
This is not the component; it's a component definition. So maybe call it component_definition or component_spec.

Also rich_text_components --> rich_text_component_definitions or rich_text_component_specs. Etc.

Reply:
Changed the code structure, PTAL!